dangerzone/container_utils.py

```diff
@@ -175,27 +189,110 @@ def load_image_tarball() -> None:
-        bad_tag = f"localhost/{expected_tag}:latest"
-        good_tag = f"{CONTAINER_NAME}:{expected_tag}"
-
-        log.debug(
-            f"Dangerzone images loaded in Podman v3.4 usually have an invalid tag."
-            " Fixing it..."
+
+def tag_image_by_digest(digest: str, tag: str) -> None:
+    """Tag a container image by digest.
+    The sha256: prefix should be omitted from the digest.
+    """
+    runtime = Runtime()
+    image_id = get_image_id_by_digest(digest)
+    cmd = [str(runtime.path), "tag", image_id, tag]
+    log.debug(" ".join(cmd))
+    subprocess_run(cmd, check=True)
+
+
+def get_image_id_by_digest(digest: str) -> str:
+    """Get an image ID from a digest.
+    The sha256: prefix should be omitted from the digest.
+    """
+    runtime = Runtime()
+    # There is a "digest" filter that you can use with
+    # "podman images -f digest:<digest>", but it's only available
+    # for podman >=4.4 (and bookworm ships 4.3)
+    # So, fallback on the json format instead
+    cmd = [
+        str(runtime.path),
+        "images",
+        "--format",
+        "json",
+    ]
+    log.debug(" ".join(cmd))
+    process = subprocess_run(cmd, check=True, capture_output=True)
+
+    images = json.loads(process.stdout.decode().strip())  # type:ignore[attr-defined]
+    filtered_images = [
+        image["Id"] for image in images if image["Digest"] == f"sha256:{digest}"
+    ]
+
+    if not filtered_images:
+        raise errors.ImageNotPresentException(
+            f"Unable to find an image with digest {digest}"
+        )
+    return filtered_images[0]
+
+
+def expected_image_name() -> str:
+    image_name_path = get_resource_path("image-name.txt")
+    return image_name_path.read_text().strip("\n")
+
+
+def container_pull(
+    image: str, manifest_digest: str, callback: Optional[Callable] = None
+) -> None:
+    """Pull a container image from a registry."""
+    runtime = Runtime()
+    cmd = [str(runtime.path), "pull", f"{image}@sha256:{manifest_digest}"]
+    process = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
+        bufsize=1,
+    )
+
+    if callback:
+        for line in process.stdout:  # type: ignore
+            callback(line)
+
+    process.wait()
+    if process.returncode != 0:
+        raise errors.ContainerPullException(
+            f"Could not pull the container image: {process.returncode}"
         )
-        add_image_tag(bad_tag, good_tag)
-        delete_image_tag(bad_tag)
 
-    log.info("Successfully installed container image")
+
+def get_local_image_digest(image: Optional[str] = None) -> str:
+    """
+    Returns a image hash from a local image name
+    """
+    expected_image = image or expected_image_name()
+    # Get the image hash from the "podman images" command.
+    # It's not possible to use "podman inspect" here as it
+    # returns the digest of the architecture-bound image
```

Comment:
In case of an airgapped installation, the digest we'll get here is the arch-bound one, right? If that's the case, does the type of the digest matter, if we recursively sign the images?

Reply:
That's an interesting question, thanks!

> In case of an airgapped installation, the digest we'll get here is the arch-bound one, right?

This is correct. In case of airgapped installations, we could use `podman inspect` as the downloaded signatures will match the architecture-bound image (digest).

> Does the type of the digest matter, if we recursively sign the images?

For "over-the-air" updates, signatures should match the envelope image (digest) instead, as that's what we're pushing / tagging with `latest` (the redirection to the architecture-bound image should not be visible to the `pull` command).

We could do that work ourselves (rather than letting `podman pull` do it for us, but I see no benefit in doing so).

Reply:
Thanks. I've done some changes in  ebdbc5e012431d215c4a354e83dc65dadb05c028, taking your comment but using a passive tone (removing we/our).